cadc-pkg-server/src/main/java/org/opencadc/pkg/server/PackageRunner.java

```diff
@@ -0,0 +1,260 @@
+import java.io.OutputStreamWriter;
+import java.io.PrintWriter;
+import java.net.URI;
+import java.net.URL;
+import java.util.Date;
+import java.util.Iterator;
+import org.apache.log4j.Logger;
+
+public abstract class PackageRunner implements JobRunner
+{
+    private static final Logger log = Logger.getLogger(PackageRunner.class);
+
+    protected Job job;
+
+    protected JobUpdater jobUpdater;
+    protected SyncOutput syncOutput;
+    protected ByteCountOutputStream counter;
+    protected WebServiceLogInfo logInfo;
+
+    public PackageRunner() 
+    {
+    }
+
+    /**
+     * Generate the package name
+     * @return String with package name.
+     */
+    protected abstract String generatePackageName();
+
+    /**
+     * Build an Iterator<PackageItem> instance, populated with whatever is requested in the Job
+     * provided.
+     * Requires an Iterator<PackageItem> class be implemented that builds the right format of
+     * ArchiveEntry (TarArchiveEntry or ZipArchiveEntry) in it's next() function.
+     * @param job - Job instance containing references to files to be added to the Iterator
+     * @param currentSubject - from the scope of the calling function. Use to derive correct file references.
+     * @return Iterator<PkgItem> instance - populated with references to the files in Job provided.
+     */
+    protected abstract Iterator<PackageItem> getItems();
+
+    @Override
+    public void setJob(Job job)
+    {
+        this.job = job;
+    }
+
+    @Override
+    public void setJobUpdater(JobUpdater ju)
+    {
+        this.jobUpdater = ju;
+    }
+
+    @Override
+    public void setSyncOutput(SyncOutput so)
+    {
+        this.syncOutput = so;
+    }
+
+    @Override
+    public void run()
+    {
+        logInfo = new JobLogInfo(job);
+        log.info(logInfo.start());
+        long start = System.currentTimeMillis();
+
+        doIt();
+
+        if (counter != null)
+        {
+            logInfo.setBytes(counter.getByteCount());
+        }
+        logInfo.setElapsedTime(System.currentTimeMillis() - start);
+        log.info(logInfo.end());
+    }
+
+    private void doIt()
+    {
+        ExecutionPhase ep;
+        TarWriter w = null;
+        try
+        {
+            ep = jobUpdater.setPhase(job.getID(), ExecutionPhase.QUEUED, ExecutionPhase.EXECUTING, new Date());
+            if ( !ExecutionPhase.EXECUTING.equals(ep) )
+            {
+                ep = jobUpdater.getPhase(job.getID());
+                log.debug(job.getID() + ": QUEUED -> EXECUTING [FAILED] -- DONE");
+                logInfo.setSuccess(false);
+                logInfo.setMessage("Could not set job phase to executing, was: " + ep);
+                return;
+            }
+            log.debug(job.getID() + ": QUEUED -> EXECUTING [OK]");
+
+            // TODO: there will need to be something here to write the
+            // Items out to a tar file.
+            Iterator<PackageItem> packageItems = getItems();
+
```

Comment:
In principle, once you address the other comments it is possible to write the output loop right here, something like:
```
while (packageItems.hasNext()) {
    PackageItem pi = packageItems.next();
    writer.write(pi);
}
```
This would show/verify that the api elements all fit together.

Reply:
I can add this as code with a comments, but as the function sits it would never work - there's no place that the TarWriter is instantiated yet.

Reply:
I believe there's enough function calls and documentation to show the Steps above in the PackageRunner doIt() function now.  Added functions to instantiate the syncOutput stream so TarWriter can be used. Remains to be seen how it acts when really used.